### ISSUE TITLE
Asynchronous producers in Kinesis

### ISSUE THREAD

**Issue by kamaljoshi** [aws/aws-sdk-ruby#1930]
### Issue description

Currently Kinesis API within the AWS Ruby SDK only supports synchronous producers. While benchmarking internally we found the latency of put to be around 8ms on average. This seems a lot and ideally an application using Kinesis a lot will this quite fast. So, I was looking for asynchronous producers on Kinesis, it doesn't seem like Ruby MRI has a library that supports that. From Kinesis docs it seems like the performant way to use that is KPL but that has interfaces available only for Java so that rules it out.

Do you guys have any suggestion on this? I was thinking of implementing a side extension to the main lib that comes with aws-sdk-kinesis where the main thread just pushes the job onto a list which has a worker thread which takes care of publishing messages to Kinesis asynchronously. The current approach is a fire and forget approach where you won't get acknowledgement of message deliveries and can lead to dropped messages in case like when Kinesis's APIs are down.

### Gem name
aws-sdk, aws-sdk-kinesis

### Version of Ruby
Ruby: 2.4/2.5

**Comment by kamaljoshi** [comment]
Yes, Something like KPL would be the best alternative. If KPL's core lib would have been in C I think it would have been easier to expose interface in MRI.

Right now, I think the two options are to do a futures like pattern with a lib like `concurrent-ruby` or the separate thread approach as mentioned.

If you guys don't have bandwidth I can take this up if we can agree on the implementation.

**Comment by conrosebraugh** [comment]
Is this something that is under active development? My organization has similar needs and could benefit from this feature.

**Comment by pchaganti** [comment]
👍
